DataFormats/ForwardDetId/src/HGCalTriggerDetId.cc

```diff
@@ -0,0 +1,167 @@
+#include "DataFormats/ForwardDetId/interface/HGCalTriggerDetId.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include <ostream>
+#include <iostream>
+
+const HGCalTriggerDetId HGCalTriggerDetId::Undefined(HGCalEE,0,0,0,0,0,0,0);
+
+HGCalTriggerDetId::HGCalTriggerDetId() : DetId() {
+}
+
+HGCalTriggerDetId::HGCalTriggerDetId(uint32_t rawid) : DetId(rawid) {
+}
+
+HGCalTriggerDetId::HGCalTriggerDetId(int subdet, int zp, int type, int layer,
+				     int waferU, int waferV, int cellU,
+				     int cellV) : DetId(HGCalTrigger,ForwardEmpty) {  
+
+  int waferUabs(std::abs(waferU)), waferVabs(std::abs(waferV));
+  int waferUsign = (waferU >= 0) ? 0 : 1;
+  int waferVsign = (waferV >= 0) ? 0 : 1;
+  int zside      = (zp < 0) ? 1 : 0;
+  id_ |= (((cellU     & kHGCalCellUMask) << kHGCalCellUOffset) |
+	  ((cellV     & kHGCalCellVMask) << kHGCalCellVOffset) |
+	  ((waferUabs & kHGCalWaferUMask) << kHGCalWaferUOffset) |
+	  ((waferUsign& kHGCalWaferUSignMask) << kHGCalWaferUSignOffset) |
+	  ((waferVabs & kHGCalWaferVMask) << kHGCalWaferVOffset) |
+	  ((waferVsign& kHGCalWaferVSignMask) << kHGCalWaferVSignOffset) |
+	  ((layer     & kHGCalLayerMask) << kHGCalLayerOffset) |
+	  ((zside     & kHGCalZsideMask) << kHGCalZsideOffset) |
+	  ((type      & kHGCalTypeMask)  << kHGCalTypeOffset) |
+	  ((subdet    & kHGCalSubdetMask)<< kHGCalSubdetOffset));
+}
+
+HGCalTriggerDetId::HGCalTriggerDetId(const DetId& gen) {
+  if (!gen.null()) {
+    if ((gen.det()!=HGCalTrigger) ||
+	(((gen.subdetId()&kHGCalSubdetMask)!=HGCalEETrigger) && 
+	 ((gen.subdetId()&kHGCalSubdetMask)!=HGCalHSiTrigger))) {
+      throw cms::Exception("Invalid DetId") << "Cannot initialize HGCalTriggerDetId from " << std::hex << gen.rawId() << std::dec; 
+    }  
+  }
+  id_ = gen.rawId();
+}
+
+HGCalTriggerDetId& HGCalTriggerDetId::operator=(const DetId& gen) {
+  if (!gen.null()) {
+    if ((gen.det()!=HGCalTrigger) ||
+	(((gen.subdetId()&kHGCalSubdetMask)!=HGCalEETrigger) && 
+	 ((gen.subdetId()&kHGCalSubdetMask)!=HGCalHSiTrigger))) {
+      throw cms::Exception("Invalid DetId") << "Cannot assign HGCalTriggerDetId from " << std::hex << gen.rawId() << std::dec; 
+    }  
+  }
+  id_ = gen.rawId();
+  return (*this);
+}
+
+int HGCalTriggerDetId::triggerCellX() const {
+  int nT = (type() == HGCSiliconDetId::HGCalFine) ? 
+    HGCSiliconDetId::HGCalFineTrigger : HGCSiliconDetId::HGCalCoarseTrigger;
+  int N  = nT*HGCalTriggerCell;
+  std::vector<int> vc = cellV();
+  int x(0);
+  for (auto const & v : vc) 
```

Comment:
@bsunanda , here and below I would use brackets to make for loops more visible. Is there any chance that cellV() may be zero length? Likely not.

Reply:
I don't think so